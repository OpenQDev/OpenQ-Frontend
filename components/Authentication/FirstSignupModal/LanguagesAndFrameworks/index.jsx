import React, { useEffect, useState, useContext } from 'react';
import AddSkill from '../AddSkill';
import StoreContext from '../../../../store/Store/StoreContext';

const LanguagesAndFrameworks = ({ user }) => {
  const [appState, appDispatch] = useContext(StoreContext);
  const { accountData } = appState;
  const languages = [
    'python',
    'java',
    'c++',
    'c#',
    'javascript',
    'ruby',
    'swift',
    'go',
    'scala',
    'kotlin',
    'perl',
    'php',
    'pascal',
    'objective-c',
    'r',
    'rust',
    'dart',
    'elixir',
    'erlang',
    'f#',
    'haskell',
    'html',
    'css',
    'clojure',
    'crystal',
    'elm',
    'groovy',
    'julia',
    'nim',
    'ocaml',
    'red',
    'typescript',
    'solidity',
    'scheme',
    'smalltalk',
    'sql',
  ];
  const frameworks = [
    'django',
    'flask',
    'rails',
    'laravel',
    'express',
    'spring',
    'asp.net',
    'react',
    'vue',
    'angular',
    'ember',
    'meteor',
    'sails',
    'sinatra',
    'express.js',
    'koa',
    'hapi',
    'socket.io',
    'next.js',
    'nuxt.js',
    'strapi',
    'prisma',
    'nestjs',
    'adonis',
    'loopback',
    'svelte',
    'tailwind',
    'bulma',
    'element-ui',
    'ant-design',
    'bootstrap',
    'material-ui',
    'semantic-ui',
  ];
  const categories = {
    languages: languages,
    frameworks: frameworks,
    otherRoles: [],
  };

  const updateApiAndState = async (newRoles, accountData, appState) => {
    const newRolesAndCategory = { ...rolesInCategories, ...newRoles };
    const userValues = accountData.github
      ? {
          github: accountData.github,
          ...newRolesAndCategory,
        }
      : {
          email: accountData.email,
          ...newRolesAndCategory,
        };
    const { updateUser } = await appState.openQPrismaClient.updateUser(userValues);
    appDispatch({ type: 'UPDATE_ACCOUNT_DATA', payload: updateUser });
    setRolesInCategories(newRolesAndCategory);
  };

  const rolesInCategoriesState = useState({
    languages: [],
    frameworks: [],
    otherRoles: [],
  });
  const [rolesInCategories, setRolesInCategories] = rolesInCategoriesState;
  const [inputValue, setInputValue] = useState('');

  const handleKeypress = (e, inputs) => {
    //it triggers by pressing the enter key
    if (e.key === 'Enter' && e.target.value != '') {
      checkCategory(inputs);
    }
  };
  useEffect(() => {
    const addUsersLanguages = async () => {
      await checkCategory(user.recentLanguages?.map((lang) => lang.toLowerCase()));
    };
    addUsersLanguages();
  }, [user]);
  const handleInputChange = (e) => {
    const { value } = e.target;
    setInputValue(value);
    // setValidUsername(true); Check if lang or framework here
  };
  const checkCategory = async (inputs) => {
    const allSkils = Object.values(rolesInCategories).flat();
    const filteredInputs = inputs.filter((input) => !allSkils.includes(input));
    if (!filteredInputs.length) return;
    let updateObj = { ...rolesInCategories };
    const createUpdateObj = (nameOfCategory, updateObj, input) => {
      updateObj = { ...updateObj, [nameOfCategory]: [...updateObj[nameOfCategory], input] };
      return updateObj;
    };
    const categoriesArr = Object.entries(categories);
    for (const input of filteredInputs) {
      let matches = false;
      for (const category of categoriesArr) {
        const nameOfCategory = category[0];
        const listOfRolesInCategory = category[1];
        if (listOfRolesInCategory.includes(input.toLowerCase())) {
          updateObj = createUpdateObj(nameOfCategory, updateObj, input);
          matches = true;
        }
      }
      if (!matches) updateObj = createUpdateObj('otherRoles', updateObj, input);
    }
    await updateApiAndState(updateObj, accountData, appState);

    setInputValue('');
  };
  return (
    <>
      <p className='font-semibold pt-4 py-2'>Tell us about your skills</p>
      <div className='flex flex-wrap'>
        {Object.entries(rolesInCategories).map((category) => {
          const nameOfCategory = category[0];
          return (
            <AddSkill
              rolesInCategoriesState={rolesInCategoriesState}
              key={category}
              category={nameOfCategory}
              setInputValue={setInputValue}
            />
          );
        })}
      </div>
      <div className='flex justify-center gap-2 mt-2'>
        <input
          className='flex input-field w-full'
          placeholder='Programming languages, frameworks...'
          onChange={handleInputChange}
          onKeyDown={(e) => handleKeypress(e, [inputValue])}
          value={inputValue}
        />
        <button onClick={() => checkCategory([inputValue])} className='btn-primary'>
          Add
        </button>
      </div>
    </>
  );
};

export default LanguagesAndFrameworks;
