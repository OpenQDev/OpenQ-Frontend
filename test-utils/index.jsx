// Third party
import React from 'react';
// Cusotm
import { render, renderHook } from '@testing-library/react';
import { cleanup } from '@testing-library/react';

import StoreProvider from '../store/Store/TestStoreProvider';
import AuthProvider from '../store/AuthStore/TestAuthProvider';
import FundProvider from '../components/FundBounty/FundStore/TestFundProvider';
import TokenProvider from '../components/TokenSelection/TokenStore/TokenProvider';

// Add in any providers here if necessary:
// test-utils.js
// (ReduxProvider, ThemeProvider, etc)
afterEach(() => {
  cleanup();
});

const customRender = (ui, options = {}, storeProps, authProps) => {
  const IntersectionObserverMock = vi.fn(() => ({
    disconnect: vi.fn(),
    observe: vi.fn(),
    takeRecords: vi.fn(),
    unobserve: vi.fn(),
  }));
  vi.stubGlobal('IntersectionObserver', IntersectionObserverMock);
  const Providers = ({ children }) => {
    return (
      <AuthProvider StoreProps={authProps}>
        <StoreProvider StoreProps={storeProps}>
          <TokenProvider>{children}</TokenProvider>
        </StoreProvider>
      </AuthProvider>
    );
  };

  return render(ui, { wrapper: Providers, ...options });
};
const customHookRender = (hook, options = {}, storeProps, authProps, fundProps) => {
  const Providers = ({ children }) => {
    return (
      <AuthProvider StoreProps={authProps}>
        <StoreProvider StoreProps={storeProps}>
          <FundProvider StoreProps={fundProps}>
            <TokenProvider>{children}</TokenProvider>
          </FundProvider>
        </StoreProvider>
      </AuthProvider>
    );
  };

  return renderHook(hook, { wrapper: Providers, ...options });
};

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render, customHookRender as hookRender };
