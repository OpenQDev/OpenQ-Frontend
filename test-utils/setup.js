import { expect, beforeEach } from 'vitest';
import { cleanup } from '@testing-library/react';
import matchers from '@testing-library/jest-dom/matchers';

expect.extend(matchers);
beforeEach(() => {
  cleanup();
});
process.on('unhandledRejection', (reason) => {
  // eslint-disable-next-line no-console
  console.log(`FAILED TO HANDLE PROMISE REJECTION`);
  throw reason;
});
