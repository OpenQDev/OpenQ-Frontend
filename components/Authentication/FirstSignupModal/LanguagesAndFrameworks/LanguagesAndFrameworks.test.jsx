import { render } from '../../../../test-utils';
import React from 'react';
import LanguagesAndFrameworks from '.';

describe('Languages and Frameworks', () => {
  const user = {
    id: '63c58870f2943bf006029305',
    renderError: '',
    avatarUrl: 'https://avatars.githubusercontent.com/u/75732239?u=285a48b76fb74f9487a3f595002bdbd69522a151&v=4',
    __typename: 'User',
    name: 'Anya',
    login: 'ArcAnya',
    websiteUrl: 'arcanya.com',
    url: 'https://github.com/ArcAnya',
    twitterUsername: null,
    recentLanguages: ['Dockerfile', 'JavaScript', 'CSS', 'TypeScript', 'Mustache'],
    onChainAddress: null,
    github: 'MDQ6VXNlcjc1NzMyMjM5',
    devRoles: [],
    otherRoles: ['ohia'],
    interests: [],
    languages: ['dockerfile', 'javascript', 'css', 'typescript', 'mustache', 'rust'],
    frameworks: [],
  };
  it('should match initial DOM Snapshot', () => {
    const { asFragment } = render(<LanguagesAndFrameworks user={user} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
