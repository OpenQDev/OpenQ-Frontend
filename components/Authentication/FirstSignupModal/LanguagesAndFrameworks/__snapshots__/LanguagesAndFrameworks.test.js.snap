// vi Snapshot v1, https://goo.gl/fbAQLP

exports[`Languages and Frameworks should match initial DOM Snapshot 1`] = `
<DocumentFragment>
  <p
    class="font-semibold pt-4 py-2"
  >
    Tell us about your skills
  </p>
  <div
    class="flex flex-wrap"
  />
  <div
    class="flex justify-center gap-2 mt-2"
  >
    <input
      class="flex input-field w-full"
      placeholder="Programming languages, frameworks..."
      value=""
    />
    <button
      class="btn-primary"
    >
      Add
    </button>
  </div>
</DocumentFragment>
`;
